File: HMCL/src/core/java/org/jackhuang/hellominecraft/launcher/core/install/optifine/bmcl/OptiFineBMCLVersionList.java
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2013  huangyuhui <huanghongxun2008@126.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hellominecraft.launcher.core.install.optifine.bmcl;

import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import org.jackhuang.hellominecraft.util.C;
import org.jackhuang.hellominecraft.util.ArrayUtils;
import org.jackhuang.hellominecraft.launcher.core.install.InstallerVersionList;
import org.jackhuang.hellominecraft.launcher.core.install.optifine.OptiFineVersion;
import org.jackhuang.hellominecraft.util.StrUtils;
import org.jackhuang.hellominecraft.util.tasks.Task;
import org.jackhuang.hellominecraft.util.tasks.TaskInfo;
import org.jackhuang.hellominecraft.util.tasks.download.HTTPGetTask;

/**
 *
 * @author huangyuhui
 */
public class OptiFineBMCLVersionList extends InstallerVersionList {

    private static OptiFineBMCLVersionList instance;

    public static OptiFineBMCLVersionList getInstance() {
        if (null == instance)
            instance = new OptiFineBMCLVersionList();
        return instance;
    }

    public ArrayList<OptiFineVersion> root;

    private static final Type TYPE = new TypeToken<ArrayList<OptiFineVersion>>() {
    }.getType();

    @Override
    public Task refresh(String[] needed) {
        return new TaskInfo(C.i18n("install.optifine.get_list")) {
            HTTPGetTask task = new HTTPGetTask("http://bmclapi.bangbang93.com/optifine/versionlist");

            @Override
            public Collection<Task> getDependTasks() {
                return Arrays.asList(task.setTag("BMCL Optifine Download Site"));
            }

            @Override
            public void executeTask(boolean areDependTasksSucceeded) throws Throwable {
                String s = task.getResult();

                versionMap = new HashMap<>();
                versions = new ArrayList<>();
                
                HashSet<String> duplicates = new HashSet<>();

                if (s == null)
                    return;
                root = C.GSON.fromJson(s, TYPE);
                for (OptiFineVersion v : root) {
                    v.setVersion(v.type + '_' + v.patch);
                    v.setMirror(String.format("http://bmclapi2.bangbang93.com/optifine/%s/%s/%s", v.getMCVersion(), v.type, v.patch));
                    if (duplicates.contains(v.getMirror()))
                        continue;
                    else
                        duplicates.add(v.getMirror());
                    InstallerVersion iv = new InstallerVersion(v.getVersion(), StrUtils.formatVersion(v.getMCVersion()));

                    List<InstallerVersion> al = ArrayUtils.tryGetMapWithList(versionMap, StrUtils.formatVersion(v.getMCVersion()));
                    iv.installer = iv.universal = v.getMirror();
                    al.add(iv);
                    versions.add(iv);
                }

                Collections.sort(versions, InstallerVersionComparator.INSTANCE);
            }
        };
    }

    @Override
    public String getName() {
        return "OptiFine - BMCLAPI(By: bangbang93)";
    }

}

File: HMCLCore/src/main/java/org/jackhuang/hmcl/download/OptiFineBMCLVersionList.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.download

import org.jackhuang.hmcl.task.GetTask
import org.jackhuang.hmcl.task.Task
import org.jackhuang.hmcl.util.GSON
import org.jackhuang.hmcl.util.asVersion
import org.jackhuang.hmcl.util.toURL
import org.jackhuang.hmcl.util.typeOf
import java.util.TreeSet

object OptiFineBMCLVersionList : VersionList<Unit>() {
    override fun refreshAsync(downloadProvider: DownloadProvider): Task {
        return RefreshTask(downloadProvider)
    }

    private class RefreshTask(val downloadProvider: DownloadProvider): Task() {
        val task = GetTask("http://bmclapi.bangbang93.com/optifine/versionlist".toURL())
        override val dependents: Collection<Task> = listOf(task)
        override fun execute() {
            versionMap.clear()
            versions.clear()

            val duplicates = mutableSetOf<String>()
            val root = GSON.fromJson<List<OptiFineVersion>>(task.result!!, typeOf<List<OptiFineVersion>>())
            for (element in root) {
                val version = element.type ?: continue
                val mirror = "http://bmclapi2.bangbang93.com/optifine/${element.gameVersion}/${element.type}/${element.patch}"
                if (duplicates.contains(mirror))
                    continue
                else
                    duplicates += mirror

                val gameVersion = element.gameVersion?.asVersion() ?: continue
                val remoteVersion = RemoteVersion<Unit>(
                        gameVersion = gameVersion,
                        selfVersion = version,
                        url = mirror,
                        tag = Unit
                )

                val set = versionMap.getOrPut(gameVersion, { TreeSet<RemoteVersion<Unit>>() }) as MutableCollection<RemoteVersion<Unit>>
                set.add(remoteVersion)
                versions.add(remoteVersion)
            }
        }

    }

}